### PR TITLE
313 activity complex element complete status

### PR DESCRIPTION
**Pull request by man-git007**

### CODE REVIEW

File: app/IATI/Services/ElementCompleteService.php [hunk 1]
```diff
@@ -0,0 +1,1197 @@
...
+                $parent = $parentLevel[$mandatoryAttribute]['parent'];
+
+                /*checks if parent attribute have specific value for child attribute to be relevant*/
+                if (!in_array($data[$parent['name']], $parent['value'], true)) {
+                    continue;
+                }
+            }
+
+            if (!array_key_exists($mandatoryAttribute, $data) || (empty($data[$mandatoryAttribute]))) {
+                //dd('isAttributeDataCompleted fx called1', ' Attribute is empty', 'attribute-check:', $mandatoryAttributes, $mandatoryAttribute, $data);
```
Comment (Sanilblank) [inline app/IATI/Services/ElementCompleteService.php:147]: remove comment code

File: app/IATI/Services/ElementCompleteService.php [hunk 2]
```diff
@@ -0,0 +1,1197 @@
...
+    {
+        $mandatoryAttributes = $this->getMandatoryAttributes($elementSchema);
+
+        if (!empty($mandatoryAttributes)) {
+            if (empty($data)) {
+                return false;
+            }
+
+            if (!$this->isAttributeDataCompleted($mandatoryAttributes, $data)) {
+                //dd('singleDimensionAttributeCheck fx called1', 'Level2 single dimension attribute is empty', 'attribute-check:', $mandatoryAttributes, $data);
```
Comment (Sanilblank) [inline app/IATI/Services/ElementCompleteService.php:175]: remove comment code

File: app/IATI/Services/ElementCompleteService.php [hunk 3]
```diff
@@ -0,0 +1,1197 @@
...
+     */
+    public function isMultiDimensionAttributeCompleted($elementSchema, $data): bool
+    {
+        $mandatoryAttributes = $this->getMandatoryAttributes($elementSchema);
+
+        if (!empty($mandatoryAttributes)) {
+            if (empty($data)) {
+                return false;
+            }
+            foreach ($data as $datum) {
```
Comment (Sanilblank) [inline app/IATI/Services/ElementCompleteService.php:201]: new line

File: app/IATI/Services/ElementCompleteService.php [hunk 4]
```diff
@@ -0,0 +1,1197 @@
...
+            if (empty($items)) {
+                //dd('isSubElementDataCompleted fx called2', 'Sub element array is empty', 'sub-element-check:', $mandatorySubElement, $data, $items);
+
+                return false;
+            }
+
+            foreach ($mandatorySubElement as $mandatoryField) {
+                foreach ($items as $item) {
+                    if (!array_key_exists($mandatoryField, $item) || (empty($item[$mandatoryField]))) {
+                        //dd('isSubElementDataCompleted fx called3', 'Sub element is empty', 'sub-element-check:', $mandatoryField, $item);
```
Comment (Sanilblank) [inline app/IATI/Services/ElementCompleteService.php:246]: remove commented code

File: app/IATI/Services/ElementCompleteService.php [hunk 5]
```diff
@@ -0,0 +1,1197 @@
...
+                    return false;
+                }
+
+                if (empty($data[$key])) {
+                    //dd('isSubElementCompleted fx is called2', 'Sub element empty', 'mandatory-sub-element_attributes:', $mandatorySubElementAttributes, 'mandatory-child-sub-elements: ',
+                    // $mandatoryChildSubElements, $key, $data);
+
+                    return false;
+                }
+                if (!$this->isElementCompleted($mandatorySubElementAttributes, $mandatoryChildSubElements, $data[$key])) {
```
Comment (Sanilblank) [inline app/IATI/Services/ElementCompleteService.php:321]: new line

File: app/IATI/Services/ElementCompleteService.php [hunk 6]
```diff
@@ -0,0 +1,1197 @@
...
+                $mandatorySubElementsFlag = true;
+                break;
+            }
+        }
+
+        if ($mandatorySubElementsFlag) {
+            if (empty($data)) {
+                return false;
+            }
+            foreach ($data as $datum) {
```
Comment (Sanilblank) [inline app/IATI/Services/ElementCompleteService.php:427]: new line

File: app/IATI/Services/ElementCompleteService.php [hunk 7]
```diff
@@ -0,0 +1,1197 @@
...
+            }
+
+            if (!array_key_exists($key, $data)) {
+                //dd('isLevelThreeSingleDimensionElementCompleted fx called1', 'sub-element-empty', 'sub-element-check:', $mandatorySubElementAttributes, $key, $data);
+
+                return false;
+            }
+
+            if (empty($data[$key])) {
+                //dd('isLevelThreeSingleDimensionElementCompleted fx called2', 'sub-element-empty', 'sub-element-check:', $mandatorySubElementAttributes, $key, $data);
```
Comment (Sanilblank) [inline app/IATI/Services/ElementCompleteService.php:481]: remove commented code

File: app/Observers/ResultObserver.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+use App\IATI\Models\Activity\Result;
+use App\IATI\Services\ElementCompleteService;
+
+/**
+ * Class ResultObserver.
+ */
+class ResultObserver
+{
+    protected ElementCompleteService $elementCompleteService;
```
Comment (Sanilblank) [inline app/Observers/ResultObserver.php:13]: doc comment

File: app/Observers/TransactionObserver.php [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+use App\IATI\Models\Activity\Transaction;
+use App\IATI\Services\ElementCompleteService;
+
+/**
+ * Class TransactionObserver.
+ */
+class TransactionObserver
+{
+    protected ElementCompleteService $elementCompleteService;
```
Comment (Sanilblank) [inline app/Observers/TransactionObserver.php:13]: doc comment

File: app/Providers/EventServiceProvider.php [hunk 10]
```diff
@@ -2,10 +2,21 @@
...
+use App\Observers\OrganizationObserver;
+use App\Observers\PeriodObserver;
+use App\Observers\ResultObserver;
+use App\Observers\TransactionObserver;
 use Illuminate\Auth\Events\Registered;
 use Illuminate\Auth\Listeners\SendEmailVerificationNotification;
 use Illuminate\Foundation\Support\Providers\EventServiceProvider as ServiceProvider;
-use Illuminate\Support\Facades\Event;
 
 class EventServiceProvider extends ServiceProvider
```
Comment (Sanilblank) [inline app/Providers/EventServiceProvider.php:21]: doc comment

File: database/factories/IATI/Models/Activity/TransactionFactory.php [hunk 11]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+namespace Database\Factories\IATI\Models\Activity;
+
+use App\IATI\Models\Activity\Transaction;
+use Illuminate\Database\Eloquent\Factories\Factory;
+
+class TransactionFactory extends Factory
```
Comment (Sanilblank) [inline database/factories/IATI/Models/Activity/TransactionFactory.php:8]: doc comment

File: database/factories/IATI/Models/Activity/TransactionFactory.php [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+{
+    protected $model = Transaction::class;
```
Comment (Sanilblank) [inline database/factories/IATI/Models/Activity/TransactionFactory.php:10]: doc comment

File: tests/Feature/Element/ElementCompleteTest.php [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+
+use App\IATI\Services\ElementCompleteService;
+use Tests\TestCase;
+
+/**
+ * Class ElementCompleteTest.
+ */
+class ElementCompleteTest extends TestCase
+{
+    protected ElementCompleteService $elementCompleteService;
```
Comment (Sanilblank) [inline tests/Feature/Element/ElementCompleteTest.php:13]: Doc comment

File: tests/Feature/Element/ReportingOrgCompleteTest.php [hunk 14]
```diff
@@ -0,0 +1,226 @@
+<?php
+
+namespace Tests\Feature\Element;
+
+/**
+ * Class ReportingOrgCompleteTest.
+ */
+class ReportingOrgCompleteTest extends ElementCompleteTest
+{
+    private string $element = 'reporting_org';
```
Comment (Sanilblank) [inline tests/Feature/Element/ReportingOrgCompleteTest.php:10]: doc comment

File: app/IATI/Services/ElementCompleteService.php [hunk 15]
```diff
@@ -0,0 +1,1197 @@
+<?php
+
```
Comment (sriza) [inline app/IATI/Services/ElementCompleteService.php:2]: declare strict type

File: app/IATI/Services/ElementCompleteService.php [hunk 16]
```diff
@@ -0,0 +1,1197 @@
...
+    public function isAttributeDataCompleted($mandatoryAttributes, $data): bool
+    {
+        if (empty($mandatoryAttributes)) {
+            return true;
+        }
+
+        if (empty($data)) {
+            return false;
+        }
+        $elementSchema = getElementSchema($this->element);
```
Comment (sriza) [inline app/IATI/Services/ElementCompleteService.php:127]: new line

File: app/IATI/Services/ElementCompleteService.php (hunk 3 above)
Comment (sriza) [inline app/IATI/Services/ElementCompleteService.php:201]: add new line

File: app/IATI/Services/ElementCompleteService.php [hunk 17]
```diff
@@ -0,0 +1,1197 @@
...
+
+            foreach ($data as $datum) {
+                if (!$this->isAttributeDataCompleted($mandatoryAttributes, $datum)) {
+                    //dd('isElementCompleted fx is called1', 'Attribute is empty', 'mandatory-attributes:', $mandatoryAttributes, $data, $datum);
+
+                    return false;
+                }
+
+                if (!$this->isSubElementDataCompleted($mandatorySubElements, $datum)) {
+                    //dd('isElementCompleted fx is called2', 'Sub element is empty', 'mandatory-sub-element:', $mandatorySubElements, $data, $datum);
```
Comment (sriza) [inline app/IATI/Services/ElementCompleteService.php:282]: remove dd

File: app/IATI/Services/ElementCompleteService.php [hunk 18]
```diff
@@ -0,0 +1,1197 @@
...
+    public function isElementCompleted($mandatoryAttributes, $mandatorySubElements, $data): bool
+    {
+        if (!empty($mandatoryAttributes) || !empty($mandatorySubElements)) {
+            if (empty($data)) {
+                return false;
+            }
+
+            foreach ($data as $datum) {
+                if (!$this->isAttributeDataCompleted($mandatoryAttributes, $datum)) {
+                    //dd('isElementCompleted fx is called1', 'Attribute is empty', 'mandatory-attributes:', $mandatoryAttributes, $data, $datum);
```
Comment (sriza) [inline app/IATI/Services/ElementCompleteService.php:276]: remove dd

File: app/IATI/Services/ElementCompleteService.php [hunk 19]
```diff
@@ -0,0 +1,1197 @@
...
+        foreach ($mandatorySubElements as $key => $mandatorySubElement) {
+            if (!array_key_exists($key, $data)) {
+                //dd('isSubElementDataCompleted fx called1', 'Whole Sub element has not filled yet', 'sub-element-check:', $mandatorySubElement, $data);
+
+                return false;
+            }
+            $items = $data[$key];
+
+            if (empty($items)) {
+                //dd('isSubElementDataCompleted fx called2', 'Sub element array is empty', 'sub-element-check:', $mandatorySubElement, $data, $items);
```
Comment (sriza) [inline app/IATI/Services/ElementCompleteService.php:238]: remove dd

File: app/Observers/ActivityObserver.php [hunk 20]
```diff
@@ -0,0 +1,76 @@
+<?php
+
```
Comment (sriza) [inline app/Observers/ActivityObserver.php:2]: declare strict type

File: app/Observers/IndicatorObserver.php [hunk 21]
```diff
@@ -0,0 +1,41 @@
+<?php
+
```
Comment (sriza) [inline app/Observers/IndicatorObserver.php:2]: declare strict type

File: app/Observers/OrganizationObserver.php [hunk 22]
```diff
@@ -0,0 +1,26 @@
+<?php
+
```
Comment (sriza) [inline app/Observers/OrganizationObserver.php:2]: declare strict type

File: app/Observers/PeriodObserver.php [hunk 23]
```diff
@@ -0,0 +1,41 @@
+<?php
+
```
Comment (sriza) [inline app/Observers/PeriodObserver.php:2]: declare strict type

File: app/Http/Controllers/Admin/Activity/ActivityController.php [hunk 24]
```diff
@@ -113,10 +113,12 @@ public function show(Activity $activity): View|JsonResponse|RedirectResponse
             $elementGroups = json_decode(file_get_contents(app_path('Data/Activity/ElementGroup.json')), true);
             $types = $this->getActivityDetailDataType();
             $status = $this->getActivityDetailStatus($activity);
-            $progress = 75;
+            $progress = $this->activityPublishingProgress($status);
+            dd($progress);
```
[the remark this answers is not in the document]
Comment (man-git007) [reply app/Http/Controllers/Admin/Activity/ActivityController.php]: has been resolved

File: app/Observers/ResultObserver.php [hunk 25]
```diff
@@ -0,0 +1,67 @@
+<?php
+
```
Comment (sriza) [inline app/Observers/ResultObserver.php:2]: declare strict type

File: app/Http/Controllers/Admin/Activity/ActivityController.php [hunk 26]
```diff
@@ -113,10 +113,12 @@ public function show(Activity $activity): View|JsonResponse|RedirectResponse
...
 
             return view('admin.activity.show', compact('elements', 'elementGroups', 'progress', 'activity', 'toast', 'types', 'status'));
         } catch (Exception $e) {
+            dd($e->getMessage());
```
[the remark this answers is not in the document]
Comment (man-git007) [reply app/Http/Controllers/Admin/Activity/ActivityController.php]: has been removed

File: app/Observers/TransactionObserver.php [hunk 27]
```diff
@@ -0,0 +1,66 @@
+<?php
+
```
Comment (sriza) [inline app/Observers/TransactionObserver.php:2]: declare strict type

File: app/IATI/Models/Activity/Activity.php [hunk 28]
```diff
@@ -16,6 +16,8 @@ class Activity extends Model
 
     protected $appends = ['title_element_completed'];
 
+    public $element = '';
```
[the remark this answers is not in the document]
Comment (man-git007) [reply app/IATI/Models/Activity/Activity.php]: have been removed

File: tests/Feature/CoreElementTest.php [hunk 29]
```diff
@@ -0,0 +1,95 @@
+<?php
+
```
Comment (sriza) [inline tests/Feature/CoreElementTest.php:2]: declare strict type

File: app/IATI/Models/Activity/Activity.php [hunk 30]
```diff
@@ -158,15 +160,15 @@ public function mandatorySubElements($fields): array
         foreach ($fields as $field) {
             $mandatoryFields = [];
 
-            if ($field['criteria'] == 'mandatory') {
+            if (array_key_exists('criteria', $field) && $field['criteria'] == 'mandatory') {
```
[the remark this answers is not in the document]
Comment (man-git007) [reply app/IATI/Models/Activity/Activity.php]: Moved to `ElementCompleteService.php` file

File: database/factories/IATI/Models/Activity/TransactionFactory.php (hunk 12 above)
Comment (man-git007) [reply database/factories/IATI/Models/Activity/TransactionFactory.php:18]: added

File: tests/Feature/Element/ElementCompleteTest.php (hunk 13 above)
Comment (man-git007) [reply tests/Feature/Element/ElementCompleteTest.php:18]: added

File: tests/Feature/Element/ReportingOrgCompleteTest.php (hunk 14 above)
Comment (man-git007) [reply tests/Feature/Element/ReportingOrgCompleteTest.php:17]: added

File: app/IATI/Services/ElementCompleteService.php (hunk 16 above)
Comment (man-git007) [reply app/IATI/Services/ElementCompleteService.php:128]: added

File: app/IATI/Services/ElementCompleteService.php (hunk 17 above)
Comment (man-git007) [reply app/IATI/Services/ElementCompleteService.php:284]: need for debugging. Ignore for now

File: app/Observers/OrganizationObserver.php (hunk 22 above)
Comment (man-git007) [reply app/Observers/OrganizationObserver.php:2]: added

File: app/Observers/ResultObserver.php (hunk 25 above)
Comment (man-git007) [reply app/Observers/ResultObserver.php:2]: added

File: tests/Feature/CoreElementTest.php (hunk 29 above)
Comment (man-git007) [reply tests/Feature/CoreElementTest.php:2]: added

File: app/IATI/Models/Activity/Activity.php [hunk 31]
```diff
@@ -16,6 +16,8 @@ class Activity extends Model
 
     protected $appends = ['title_element_completed'];
```
[the remark this answers is not in the document]
Comment (man-git007) [reply app/IATI/Models/Activity/Activity.php]: removed

File: app/IATI/Services/ElementCompleteService.php (hunk 15 above)
Comment (man-git007) [reply app/IATI/Services/ElementCompleteService.php:2]: added

File: app/IATI/Services/ElementCompleteService.php (hunk 18 above)
Comment (man-git007) [reply app/IATI/Services/ElementCompleteService.php:285]: ignore for now

File: app/IATI/Services/ElementCompleteService.php (hunk 3 above)
Comment (man-git007) [reply app/IATI/Services/ElementCompleteService.php:210]: added

File: app/Observers/TransactionObserver.php (hunk 27 above)
Comment (man-git007) [reply app/Observers/TransactionObserver.php:2]: added

File: app/Observers/PeriodObserver.php (hunk 23 above)
Comment (man-git007) [reply app/Observers/PeriodObserver.php:2]: added

File: app/Providers/EventServiceProvider.php (hunk 10 above)
Comment (man-git007) [reply app/Providers/EventServiceProvider.php:21]: added

**Pull request merged**